File: cpp/src/strategies/minmax/src/minmax/BoardEvaluation.cpp
#include "minmax/BoardEvaluation.hpp"
#include "minmax/config.hpp"

#include <game/Game.hpp>
#include <game/Piece.hpp>
#include <algorithm>
#include <util/Log.hpp>

namespace Alphalcazar::Strategy::MinMax {
	/*!
	 * \brief Returns a score multiplier for the heuristic score of a piece given its position and direction on the board.
	 *
	 * Grants a higher multiplier for pieces that are well positioned (ex. in the center) or have a long expected
	 * lifetime (ex. facing towards the board interior) and a lower multiplier for pieces that are badly positioned or
	 * with a short expected lifetime (ex. about to exit the board).
	 */
	float GetPieceScoreMultiplier(const Game::Coordinates& coordinates, Game::Direction direction) {
		if (coordinates.IsCenter()) {
			return c_CenterPieceMultiplier;
		}
		
		if (coordinates.IsBoardCorner()) {
			// In the board corners, a piece can only have recently entered the board
			// or be about to leave it
			const auto pieceTargetCoordinate = coordinates.GetCoordinateInDirection(direction, 1);
			if (pieceTargetCoordinate.IsPerimeter()) {
				return c_PieceAboutToExitMultiplier;
			}
			return c_FreshCornerPieceMultiplier;
		}
		
		// The piece is on one of the center lanes, but not in the center tile
		const auto pieceTargetCoordinate = coordinates.GetCoordinateInDirection(direction, 1);
		if (pieceTargetCoordinate.IsPerimeter()) {
			// The piece is about to exit the board
			return c_PieceAboutToExitMultiplier;
		}
		
		if (pieceTargetCoordinate.IsCenter()) {
			// The piece just entered the center lane and wants to move to the center
			return c_FreshCenterLanePieceMultiplier;
		}
		return 1.f;
	}

	Score GameResultToScore(Game::PlayerId playerId, Game::GameResult result) {
		switch (result) {
		case Game::GameResult::PLAYER_ONE_WINS:
			return playerId == Game::PlayerId::PLAYER_ONE ? c_WinConditionScore : -c_WinConditionScore;
		case Game::GameResult::PLAYER_TWO_WINS:
			return playerId == Game::PlayerId::PLAYER_TWO ? c_WinConditionScore : -c_WinConditionScore;
		case Game::GameResult::DRAW:
			return 0;
		default:
			Utils::LogError("Tried to convert invalid game result to score");
		}
		return 0;
	}

	Score EvaluateBoard(Game::PlayerId playerId, const Game::Game& game) {
		Score totalScore = 0;
		auto [pieces, piecesCount] = game.GetBoard().GetPieces();
		for (std::size_t i = 0; i < piecesCount; ++i) {
			auto& [coordinates, piece] = pieces[i];
			if (!coordinates.IsPerimeter()) {
				const Game::Direction direction = piece.GetMovementDirection();
				const float pieceScoreMultiplier = GetPieceScoreMultiplier(coordinates, direction);

				// The piece on board score array stores all piece types in order, meaning that
				// the score of a certain piece type will be located at index (type - 1)
				const Score pieceOnBoardScore = c_PieceOnBoardScores[piece.GetType() - 1];
				const Score pieceScore = static_cast<Score>(pieceOnBoardScore * pieceScoreMultiplier);

				// Add the score if the piece belongs to the player for which we are evaluating the score
				// or subtract it if it belongs to the opponent
				if (piece.GetOwner() == playerId) {
					totalScore += pieceScore;
				} else {
					totalScore -= pieceScore;
				}
			}
		}
		return totalScore;
	}

	Score GetDepthAdjustedScore(Score score, Depth depth) {
		const Score depthPenalty = depth * c_DepthScorePenalty;
		const Score penalty = std::min(depthPenalty, std::abs(score));
		if (score > 0) {
			score -= penalty;
		} else {
			score += penalty;
		}
		return score;
	}
}

File: strategies/tree_search/board_evaluation.py
from game.enums import GameResult
from game.board import Board
from strategies.tree_search.config import WIN_CONDITION_SCORE, TILE_SCORE_MULTIPLIER, PLACED_PIECE_SCORE, DEPTH_PENALTY

BOARD_SCORE_CACHE = dict()


def get_depth_adjusted_score(score: int) -> int:
    capped_offset = DEPTH_PENALTY if DEPTH_PENALTY < abs(score) else abs(score)
    if score > 0:
        score -= capped_offset
    else:
        score += capped_offset
    return score


def game_result_to_score(result: GameResult) -> int:
    if result == GameResult.win:
        return WIN_CONDITION_SCORE
    elif result == GameResult.draw:
        return 0
    elif result == GameResult.loss:
        return -WIN_CONDITION_SCORE


def evaluate_board(board: Board, player_id: int, opponent_id: int) -> int:
    # First check if we have this score in the cache already #
    cache_key = f"{player_id}#{board.to_string_notation()}"
    if BOARD_SCORE_CACHE.get(cache_key):
        return BOARD_SCORE_CACHE[cache_key]

    result = board.get_game_result(player_id, opponent_id)
    score = 0
    if result is not None:
        score = game_result_to_score(result)
    else:
        pieces = board.get_board_pieces(exclude_perimeter=True)
        for piece in pieces:
            tile_multiplier = TILE_SCORE_MULTIPLIER[piece.tile.x][piece.tile.y][piece.direction]
            piece_score = PLACED_PIECE_SCORE[piece.type] * tile_multiplier
            if piece.owner_id == player_id:
                score += piece_score
            else:
                score -= piece_score
    BOARD_SCORE_CACHE[cache_key] = score
    return score
